src/rate-limiter.ts

```diff
@@ -1,17 +1,77 @@
+import { CAPTURE_RATE_LIMIT } from './constants'
+import type { PostHog } from './posthog-core'
 import { RequestResponse } from './types'
 import { logger } from './utils/logger'
 
-const oneMinuteInMilliseconds = 60 * 1000
+const ONE_MINUTE_IN_MILLISECONDS = 60 * 1000
+const RATE_LIMIT_EVENT = '$$client_ingestion_warning'
 
 interface CaptureResponse {
     quota_limited?: string[]
 }
 
 export class RateLimiter {
-    limits: Record<string, number> = {}
+    instance: PostHog
+    serverLimits: Record<string, number> = {}
 
-    public isRateLimited(batchKey: string | undefined): boolean {
-        const retryAfter = this.limits[batchKey || 'events'] || false
+    captureEventsPerSecond: number
+    captureEventsBurstLimit: number
+    lastEventRateLimited = false
+
+    constructor(instance: PostHog) {
+        this.instance = instance
+
+        this.captureEventsPerSecond = instance.config.rate_limiting?.events_per_second || 10
+        this.captureEventsBurstLimit = Math.max(
+            instance.config.rate_limiting?.events_burst_limit || this.captureEventsPerSecond * 10,
+            this.captureEventsPerSecond
+        )
+
+        this.lastEventRateLimited = this.isCaptureClientSideRateLimited(true)
+    }
+
+    public isCaptureClientSideRateLimited(checkOnly = false): boolean {
+        // This is primarily to prevent runaway loops from flooding capture with millions of events for a single user.
+        // It's as much for our protection as theirs.
+        const now = new Date().getTime()
+        const bucket = this.instance.persistence?.get_property(CAPTURE_RATE_LIMIT) ?? {
+            tokens: this.captureEventsBurstLimit,
+            last: now,
+        }
+
+        bucket.tokens += ((now - bucket.last) / 1000) * this.captureEventsPerSecond
+        bucket.last = now
+
+        if (bucket.tokens > this.captureEventsBurstLimit) {
+            bucket.tokens = this.captureEventsBurstLimit
+        }
+
+        const isRateLimited = bucket.tokens < 1
+
+        if (!isRateLimited && !checkOnly) {
+            bucket.tokens = Math.max(0, bucket.tokens - 1)
+        }
+
+        if (isRateLimited && !this.lastEventRateLimited && !checkOnly) {
+            this.instance.capture(
+                RATE_LIMIT_EVENT,
```

Comment:
I worry about this DDOSing analytics ingestion (maybe not with this PR, but someone adding another warning in the future, recall how $errors DDOSed session recordings), so to be safe, could we send these events to a different ingress endpoint, e.g. `/ingestion_warning`, so we can drop it at ingress (& just point it to analytics ingestion for now, giving us flexibility to point it at a different lane in the future easily)?

Reply:
Happy for someone to do this from pipeline team

Reply:
ingress is rather infra than pipeline 🤔

Reply:
Feels like a pipeline feature even if it would require support from infra